snuba/utils/bucket_timer.py

```diff
@@ -0,0 +1,95 @@
+from __future__ import annotations
+
+from collections import defaultdict, deque
+from dataclasses import dataclass
+from datetime import datetime, timedelta
+from typing import Deque, List
+
+from snuba import settings, state
+
+
+def get_time_sum(group: List[timedelta]) -> timedelta:
+    return sum([processing_time for processing_time in group], timedelta())
+
+
+def floor_minute(time: datetime) -> datetime:
+    return time - timedelta(seconds=time.second, microseconds=time.microsecond)
+
+
+def ceil_minute(time: datetime) -> datetime:
+    if time.second == 0 and time.microsecond == 0:
+        return time
+    return floor_minute(time + timedelta(minutes=1))
+
+
+@dataclass
+class Bucket:
+    project_id: int
+    minute: datetime
+    processing_time: timedelta = timedelta(seconds=0)
+
+
+class Counter:
+    """
+    The Counter class is used to track time spent on some activity (e.g. processing a replacement) for a project.
+    To accomplish this, the `record_time_spent()` function captures some processing time range and splits it by a per
+    minute resolution (Bucket). The buckets older than settings.COUNTER_WINDOW_SIZE are trimmed. Finally, the `get_bucket_totals_exceeding_limit()`
+    function returns all project ids who's total processing time has exceeded self.limit.
+    """
+
+    def __init__(self, consumer_group: str) -> None:
+        self.consumer_group: str = consumer_group
+        self.buckets: Deque[Bucket] = deque()
+
+        percentage = state.get_config("counter_window_size_percentage", 1.0)
+        assert isinstance(percentage, float)
+        self.limit = settings.COUNTER_WINDOW_SIZE * percentage
+
+    def __trim_expired_buckets(self, now: datetime) -> None:
+        current_minute = floor_minute(now)
+        window_start = current_minute - settings.COUNTER_WINDOW_SIZE
+        while self.buckets and self.buckets[0].minute < window_start:
+            self.buckets.popleft()
+
+    def __add_to_bucket(
+        self,
+        project_id: int,
+        start_minute: datetime,
+        processing_time: timedelta,
+    ) -> None:
+        curr_bucket = None
+        for bucket in self.buckets:
+            if bucket.project_id == project_id and bucket.minute == start_minute:
```

Comment:
Makes sense, extending this idea further, I would propose that self.__buckets can even be a nested dictionary to leverage quicker access by whole minute AND project_id.

```
self.__buckets: MutableMapping[datetime, MutableMapping[int, timedelta]] = {}
```

If we let N be the total number of minutes saved and M be the total number of project_ids. Then...
For recording a bucket (creating new entry or adding processing time to existing bucket):
* Nested dictionary: O(1)
* Dictionary with deque: O(N) 

For trimming buckets:
* Nested dictionary: O(N)
* Dictionary with deque: O(N*M) <-- in most cases N will be smaller though because the probability of needing to trim the entire window is low